// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/annotations.proto

package com.google.genomics.v1;

public interface AnnotationOrBuilder extends
    // @@protoc_insertion_point(interface_extends:google.genomics.v1.Annotation)
    com.google.protobuf.MessageOrBuilder {

  /**
   * <pre>
   * The server-generated annotation ID, unique across all annotations.
   * </pre>
   *
   * <code>string id = 1;</code>
   */
  java.lang.String getId();
  /**
   * <pre>
   * The server-generated annotation ID, unique across all annotations.
   * </pre>
   *
   * <code>string id = 1;</code>
   */
  com.google.protobuf.ByteString
      getIdBytes();

  /**
   * <pre>
   * The annotation set to which this annotation belongs.
   * </pre>
   *
   * <code>string annotation_set_id = 2;</code>
   */
  java.lang.String getAnnotationSetId();
  /**
   * <pre>
   * The annotation set to which this annotation belongs.
   * </pre>
   *
   * <code>string annotation_set_id = 2;</code>
   */
  com.google.protobuf.ByteString
      getAnnotationSetIdBytes();

  /**
   * <pre>
   * The display name of this annotation.
   * </pre>
   *
   * <code>string name = 3;</code>
   */
  java.lang.String getName();
  /**
   * <pre>
   * The display name of this annotation.
   * </pre>
   *
   * <code>string name = 3;</code>
   */
  com.google.protobuf.ByteString
      getNameBytes();

  /**
   * <pre>
   * The ID of the Google Genomics reference associated with this range.
   * </pre>
   *
   * <code>string reference_id = 4;</code>
   */
  java.lang.String getReferenceId();
  /**
   * <pre>
   * The ID of the Google Genomics reference associated with this range.
   * </pre>
   *
   * <code>string reference_id = 4;</code>
   */
  com.google.protobuf.ByteString
      getReferenceIdBytes();

  /**
   * <pre>
   * The display name corresponding to the reference specified by
   * `referenceId`, for example `chr1`, `1`, or `chrX`.
   * </pre>
   *
   * <code>string reference_name = 5;</code>
   */
  java.lang.String getReferenceName();
  /**
   * <pre>
   * The display name corresponding to the reference specified by
   * `referenceId`, for example `chr1`, `1`, or `chrX`.
   * </pre>
   *
   * <code>string reference_name = 5;</code>
   */
  com.google.protobuf.ByteString
      getReferenceNameBytes();

  /**
   * <pre>
   * The start position of the range on the reference, 0-based inclusive.
   * </pre>
   *
   * <code>int64 start = 6;</code>
   */
  long getStart();

  /**
   * <pre>
   * The end position of the range on the reference, 0-based exclusive.
   * </pre>
   *
   * <code>int64 end = 7;</code>
   */
  long getEnd();

  /**
   * <pre>
   * Whether this range refers to the reverse strand, as opposed to the forward
   * strand. Note that regardless of this field, the start/end position of the
   * range always refer to the forward strand.
   * </pre>
   *
   * <code>bool reverse_strand = 8;</code>
   */
  boolean getReverseStrand();

  /**
   * <pre>
   * The data type for this annotation. Must match the containing annotation
   * set's type.
   * </pre>
   *
   * <code>.google.genomics.v1.AnnotationType type = 9;</code>
   */
  int getTypeValue();
  /**
   * <pre>
   * The data type for this annotation. Must match the containing annotation
   * set's type.
   * </pre>
   *
   * <code>.google.genomics.v1.AnnotationType type = 9;</code>
   */
  com.google.genomics.v1.AnnotationType getType();

  /**
   * <pre>
   * A variant annotation, which describes the effect of a variant on the
   * genome, the coding sequence, and/or higher level consequences at the
   * organism level e.g. pathogenicity. This field is only set for annotations
   * of type `VARIANT`.
   * </pre>
   *
   * <code>.google.genomics.v1.VariantAnnotation variant = 10;</code>
   */
  com.google.genomics.v1.VariantAnnotation getVariant();
  /**
   * <pre>
   * A variant annotation, which describes the effect of a variant on the
   * genome, the coding sequence, and/or higher level consequences at the
   * organism level e.g. pathogenicity. This field is only set for annotations
   * of type `VARIANT`.
   * </pre>
   *
   * <code>.google.genomics.v1.VariantAnnotation variant = 10;</code>
   */
  com.google.genomics.v1.VariantAnnotationOrBuilder getVariantOrBuilder();

  /**
   * <pre>
   * A transcript value represents the assertion that a particular region of
   * the reference genome may be transcribed as RNA. An alternative splicing
   * pattern would be represented as a separate transcript object. This field
   * is only set for annotations of type `TRANSCRIPT`.
   * </pre>
   *
   * <code>.google.genomics.v1.Transcript transcript = 11;</code>
   */
  com.google.genomics.v1.Transcript getTranscript();
  /**
   * <pre>
   * A transcript value represents the assertion that a particular region of
   * the reference genome may be transcribed as RNA. An alternative splicing
   * pattern would be represented as a separate transcript object. This field
   * is only set for annotations of type `TRANSCRIPT`.
   * </pre>
   *
   * <code>.google.genomics.v1.Transcript transcript = 11;</code>
   */
  com.google.genomics.v1.TranscriptOrBuilder getTranscriptOrBuilder();

  /**
   * <pre>
   * A map of additional read alignment information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */
  int getInfoCount();
  /**
   * <pre>
   * A map of additional read alignment information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */
  boolean containsInfo(
      java.lang.String key);
  /**
   * Use {@link #getInfoMap()} instead.
   */
  @java.lang.Deprecated
  java.util.Map<java.lang.String, com.google.protobuf.ListValue>
  getInfo();
  /**
   * <pre>
   * A map of additional read alignment information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */
  java.util.Map<java.lang.String, com.google.protobuf.ListValue>
  getInfoMap();
  /**
   * <pre>
   * A map of additional read alignment information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */

  com.google.protobuf.ListValue getInfoOrDefault(
      java.lang.String key,
      com.google.protobuf.ListValue defaultValue);
  /**
   * <pre>
   * A map of additional read alignment information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */

  com.google.protobuf.ListValue getInfoOrThrow(
      java.lang.String key);

  public com.google.genomics.v1.Annotation.ValueCase getValueCase();
}
