// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/references.proto

package com.google.genomics.v1;

public interface ReferenceOrBuilder extends
    // @@protoc_insertion_point(interface_extends:google.genomics.v1.Reference)
    com.google.protobuf.MessageOrBuilder {

  /**
   * <pre>
   * The server-generated reference ID, unique across all references.
   * </pre>
   *
   * <code>string id = 1;</code>
   */
  java.lang.String getId();
  /**
   * <pre>
   * The server-generated reference ID, unique across all references.
   * </pre>
   *
   * <code>string id = 1;</code>
   */
  com.google.protobuf.ByteString
      getIdBytes();

  /**
   * <pre>
   * The length of this reference's sequence.
   * </pre>
   *
   * <code>int64 length = 2;</code>
   */
  long getLength();

  /**
   * <pre>
   * MD5 of the upper-case sequence excluding all whitespace characters (this
   * is equivalent to SQ:M5 in SAM). This value is represented in lower case
   * hexadecimal format.
   * </pre>
   *
   * <code>string md5checksum = 3;</code>
   */
  java.lang.String getMd5Checksum();
  /**
   * <pre>
   * MD5 of the upper-case sequence excluding all whitespace characters (this
   * is equivalent to SQ:M5 in SAM). This value is represented in lower case
   * hexadecimal format.
   * </pre>
   *
   * <code>string md5checksum = 3;</code>
   */
  com.google.protobuf.ByteString
      getMd5ChecksumBytes();

  /**
   * <pre>
   * The name of this reference, for example `22`.
   * </pre>
   *
   * <code>string name = 4;</code>
   */
  java.lang.String getName();
  /**
   * <pre>
   * The name of this reference, for example `22`.
   * </pre>
   *
   * <code>string name = 4;</code>
   */
  com.google.protobuf.ByteString
      getNameBytes();

  /**
   * <pre>
   * The URI from which the sequence was obtained. Typically specifies a FASTA
   * format file.
   * </pre>
   *
   * <code>string source_uri = 5;</code>
   */
  java.lang.String getSourceUri();
  /**
   * <pre>
   * The URI from which the sequence was obtained. Typically specifies a FASTA
   * format file.
   * </pre>
   *
   * <code>string source_uri = 5;</code>
   */
  com.google.protobuf.ByteString
      getSourceUriBytes();

  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   */
  java.util.List<java.lang.String>
      getSourceAccessionsList();
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   */
  int getSourceAccessionsCount();
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   */
  java.lang.String getSourceAccessions(int index);
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   */
  com.google.protobuf.ByteString
      getSourceAccessionsBytes(int index);

  /**
   * <pre>
   * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
   * </pre>
   *
   * <code>int32 ncbi_taxon_id = 7;</code>
   */
  int getNcbiTaxonId();
}
