// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/references.proto

package com.google.genomics.v1;

/**
 * <pre>
 * A reference is a canonical assembled DNA sequence, intended to act as a
 * reference coordinate space for other genomic annotations. A single reference
 * might represent the human chromosome 1 or mitochandrial DNA, for instance. A
 * reference belongs to one or more reference sets.
 * For more genomics resource definitions, see [Fundamentals of Google
 * Genomics](https://cloud.google.com/genomics/fundamentals-of-google-genomics)
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.Reference}
 */
public  final class Reference extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.Reference)
    ReferenceOrBuilder {
  // Use Reference.newBuilder() to construct.
  private Reference(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private Reference() {
    id_ = "";
    length_ = 0L;
    md5Checksum_ = "";
    name_ = "";
    sourceUri_ = "";
    sourceAccessions_ = com.google.protobuf.LazyStringArrayList.EMPTY;
    ncbiTaxonId_ = 0;
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
  }
  private Reference(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    this();
    int mutable_bitField0_ = 0;
    try {
      boolean done = false;
      while (!done) {
        int tag = input.readTag();
        switch (tag) {
          case 0:
            done = true;
            break;
          default: {
            if (!input.skipField(tag)) {
              done = true;
            }
            break;
          }
          case 10: {
            java.lang.String s = input.readStringRequireUtf8();

            id_ = s;
            break;
          }
          case 16: {

            length_ = input.readInt64();
            break;
          }
          case 26: {
            java.lang.String s = input.readStringRequireUtf8();

            md5Checksum_ = s;
            break;
          }
          case 34: {
            java.lang.String s = input.readStringRequireUtf8();

            name_ = s;
            break;
          }
          case 42: {
            java.lang.String s = input.readStringRequireUtf8();

            sourceUri_ = s;
            break;
          }
          case 50: {
            java.lang.String s = input.readStringRequireUtf8();
            if (!((mutable_bitField0_ & 0x00000020) == 0x00000020)) {
              sourceAccessions_ = new com.google.protobuf.LazyStringArrayList();
              mutable_bitField0_ |= 0x00000020;
            }
            sourceAccessions_.add(s);
            break;
          }
          case 56: {

            ncbiTaxonId_ = input.readInt32();
            break;
          }
        }
      }
    } catch (com.google.protobuf.InvalidProtocolBufferException e) {
      throw e.setUnfinishedMessage(this);
    } catch (java.io.IOException e) {
      throw new com.google.protobuf.InvalidProtocolBufferException(
          e).setUnfinishedMessage(this);
    } finally {
      if (((mutable_bitField0_ & 0x00000020) == 0x00000020)) {
        sourceAccessions_ = sourceAccessions_.getUnmodifiableView();
      }
      makeExtensionsImmutable();
    }
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_descriptor;
  }

  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.Reference.class, com.google.genomics.v1.Reference.Builder.class);
  }

  private int bitField0_;
  public static final int ID_FIELD_NUMBER = 1;
  private volatile java.lang.Object id_;
  /**
   * <pre>
   * The server-generated reference ID, unique across all references.
   * </pre>
   *
   * <code>string id = 1;</code>
   */
  public java.lang.String getId() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      id_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The server-generated reference ID, unique across all references.
   * </pre>
   *
   * <code>string id = 1;</code>
   */
  public com.google.protobuf.ByteString
      getIdBytes() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      id_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int LENGTH_FIELD_NUMBER = 2;
  private long length_;
  /**
   * <pre>
   * The length of this reference's sequence.
   * </pre>
   *
   * <code>int64 length = 2;</code>
   */
  public long getLength() {
    return length_;
  }

  public static final int MD5CHECKSUM_FIELD_NUMBER = 3;
  private volatile java.lang.Object md5Checksum_;
  /**
   * <pre>
   * MD5 of the upper-case sequence excluding all whitespace characters (this
   * is equivalent to SQ:M5 in SAM). This value is represented in lower case
   * hexadecimal format.
   * </pre>
   *
   * <code>string md5checksum = 3;</code>
   */
  public java.lang.String getMd5Checksum() {
    java.lang.Object ref = md5Checksum_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      md5Checksum_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * MD5 of the upper-case sequence excluding all whitespace characters (this
   * is equivalent to SQ:M5 in SAM). This value is represented in lower case
   * hexadecimal format.
   * </pre>
   *
   * <code>string md5checksum = 3;</code>
   */
  public com.google.protobuf.ByteString
      getMd5ChecksumBytes() {
    java.lang.Object ref = md5Checksum_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      md5Checksum_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int NAME_FIELD_NUMBER = 4;
  private volatile java.lang.Object name_;
  /**
   * <pre>
   * The name of this reference, for example `22`.
   * </pre>
   *
   * <code>string name = 4;</code>
   */
  public java.lang.String getName() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      name_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The name of this reference, for example `22`.
   * </pre>
   *
   * <code>string name = 4;</code>
   */
  public com.google.protobuf.ByteString
      getNameBytes() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      name_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int SOURCE_URI_FIELD_NUMBER = 5;
  private volatile java.lang.Object sourceUri_;
  /**
   * <pre>
   * The URI from which the sequence was obtained. Typically specifies a FASTA
   * format file.
   * </pre>
   *
   * <code>string source_uri = 5;</code>
   */
  public java.lang.String getSourceUri() {
    java.lang.Object ref = sourceUri_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      sourceUri_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The URI from which the sequence was obtained. Typically specifies a FASTA
   * format file.
   * </pre>
   *
   * <code>string source_uri = 5;</code>
   */
  public com.google.protobuf.ByteString
      getSourceUriBytes() {
    java.lang.Object ref = sourceUri_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      sourceUri_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int SOURCE_ACCESSIONS_FIELD_NUMBER = 6;
  private com.google.protobuf.LazyStringList sourceAccessions_;
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   */
  public com.google.protobuf.ProtocolStringList
      getSourceAccessionsList() {
    return sourceAccessions_;
  }
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   */
  public int getSourceAccessionsCount() {
    return sourceAccessions_.size();
  }
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   */
  public java.lang.String getSourceAccessions(int index) {
    return sourceAccessions_.get(index);
  }
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   */
  public com.google.protobuf.ByteString
      getSourceAccessionsBytes(int index) {
    return sourceAccessions_.getByteString(index);
  }

  public static final int NCBI_TAXON_ID_FIELD_NUMBER = 7;
  private int ncbiTaxonId_;
  /**
   * <pre>
   * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
   * </pre>
   *
   * <code>int32 ncbi_taxon_id = 7;</code>
   */
  public int getNcbiTaxonId() {
    return ncbiTaxonId_;
  }

  private byte memoizedIsInitialized = -1;
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    if (!getIdBytes().isEmpty()) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 1, id_);
    }
    if (length_ != 0L) {
      output.writeInt64(2, length_);
    }
    if (!getMd5ChecksumBytes().isEmpty()) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 3, md5Checksum_);
    }
    if (!getNameBytes().isEmpty()) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 4, name_);
    }
    if (!getSourceUriBytes().isEmpty()) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 5, sourceUri_);
    }
    for (int i = 0; i < sourceAccessions_.size(); i++) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 6, sourceAccessions_.getRaw(i));
    }
    if (ncbiTaxonId_ != 0) {
      output.writeInt32(7, ncbiTaxonId_);
    }
  }

  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    if (!getIdBytes().isEmpty()) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(1, id_);
    }
    if (length_ != 0L) {
      size += com.google.protobuf.CodedOutputStream
        .computeInt64Size(2, length_);
    }
    if (!getMd5ChecksumBytes().isEmpty()) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(3, md5Checksum_);
    }
    if (!getNameBytes().isEmpty()) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(4, name_);
    }
    if (!getSourceUriBytes().isEmpty()) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(5, sourceUri_);
    }
    {
      int dataSize = 0;
      for (int i = 0; i < sourceAccessions_.size(); i++) {
        dataSize += computeStringSizeNoTag(sourceAccessions_.getRaw(i));
      }
      size += dataSize;
      size += 1 * getSourceAccessionsList().size();
    }
    if (ncbiTaxonId_ != 0) {
      size += com.google.protobuf.CodedOutputStream
        .computeInt32Size(7, ncbiTaxonId_);
    }
    memoizedSize = size;
    return size;
  }

  private static final long serialVersionUID = 0L;
  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.Reference)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.Reference other = (com.google.genomics.v1.Reference) obj;

    boolean result = true;
    result = result && getId()
        .equals(other.getId());
    result = result && (getLength()
        == other.getLength());
    result = result && getMd5Checksum()
        .equals(other.getMd5Checksum());
    result = result && getName()
        .equals(other.getName());
    result = result && getSourceUri()
        .equals(other.getSourceUri());
    result = result && getSourceAccessionsList()
        .equals(other.getSourceAccessionsList());
    result = result && (getNcbiTaxonId()
        == other.getNcbiTaxonId());
    return result;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    hash = (37 * hash) + ID_FIELD_NUMBER;
    hash = (53 * hash) + getId().hashCode();
    hash = (37 * hash) + LENGTH_FIELD_NUMBER;
    hash = (53 * hash) + com.google.protobuf.Internal.hashLong(
        getLength());
    hash = (37 * hash) + MD5CHECKSUM_FIELD_NUMBER;
    hash = (53 * hash) + getMd5Checksum().hashCode();
    hash = (37 * hash) + NAME_FIELD_NUMBER;
    hash = (53 * hash) + getName().hashCode();
    hash = (37 * hash) + SOURCE_URI_FIELD_NUMBER;
    hash = (53 * hash) + getSourceUri().hashCode();
    if (getSourceAccessionsCount() > 0) {
      hash = (37 * hash) + SOURCE_ACCESSIONS_FIELD_NUMBER;
      hash = (53 * hash) + getSourceAccessionsList().hashCode();
    }
    hash = (37 * hash) + NCBI_TAXON_ID_FIELD_NUMBER;
    hash = (53 * hash) + getNcbiTaxonId();
    hash = (29 * hash) + unknownFields.hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.Reference parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.Reference parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.Reference prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * A reference is a canonical assembled DNA sequence, intended to act as a
   * reference coordinate space for other genomic annotations. A single reference
   * might represent the human chromosome 1 or mitochandrial DNA, for instance. A
   * reference belongs to one or more reference sets.
   * For more genomics resource definitions, see [Fundamentals of Google
   * Genomics](https://cloud.google.com/genomics/fundamentals-of-google-genomics)
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.Reference}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.Reference)
      com.google.genomics.v1.ReferenceOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_descriptor;
    }

    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.Reference.class, com.google.genomics.v1.Reference.Builder.class);
    }

    // Construct using com.google.genomics.v1.Reference.newBuilder()
    private Builder() {
      maybeForceBuilderInitialization();
    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);
      maybeForceBuilderInitialization();
    }
    private void maybeForceBuilderInitialization() {
      if (com.google.protobuf.GeneratedMessageV3
              .alwaysUseFieldBuilders) {
      }
    }
    public Builder clear() {
      super.clear();
      id_ = "";

      length_ = 0L;

      md5Checksum_ = "";

      name_ = "";

      sourceUri_ = "";

      sourceAccessions_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      bitField0_ = (bitField0_ & ~0x00000020);
      ncbiTaxonId_ = 0;

      return this;
    }

    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_descriptor;
    }

    public com.google.genomics.v1.Reference getDefaultInstanceForType() {
      return com.google.genomics.v1.Reference.getDefaultInstance();
    }

    public com.google.genomics.v1.Reference build() {
      com.google.genomics.v1.Reference result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    public com.google.genomics.v1.Reference buildPartial() {
      com.google.genomics.v1.Reference result = new com.google.genomics.v1.Reference(this);
      int from_bitField0_ = bitField0_;
      int to_bitField0_ = 0;
      result.id_ = id_;
      result.length_ = length_;
      result.md5Checksum_ = md5Checksum_;
      result.name_ = name_;
      result.sourceUri_ = sourceUri_;
      if (((bitField0_ & 0x00000020) == 0x00000020)) {
        sourceAccessions_ = sourceAccessions_.getUnmodifiableView();
        bitField0_ = (bitField0_ & ~0x00000020);
      }
      result.sourceAccessions_ = sourceAccessions_;
      result.ncbiTaxonId_ = ncbiTaxonId_;
      result.bitField0_ = to_bitField0_;
      onBuilt();
      return result;
    }

    public Builder clone() {
      return (Builder) super.clone();
    }
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        Object value) {
      return (Builder) super.setField(field, value);
    }
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return (Builder) super.clearField(field);
    }
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return (Builder) super.clearOneof(oneof);
    }
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, Object value) {
      return (Builder) super.setRepeatedField(field, index, value);
    }
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        Object value) {
      return (Builder) super.addRepeatedField(field, value);
    }
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.Reference) {
        return mergeFrom((com.google.genomics.v1.Reference)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.Reference other) {
      if (other == com.google.genomics.v1.Reference.getDefaultInstance()) return this;
      if (!other.getId().isEmpty()) {
        id_ = other.id_;
        onChanged();
      }
      if (other.getLength() != 0L) {
        setLength(other.getLength());
      }
      if (!other.getMd5Checksum().isEmpty()) {
        md5Checksum_ = other.md5Checksum_;
        onChanged();
      }
      if (!other.getName().isEmpty()) {
        name_ = other.name_;
        onChanged();
      }
      if (!other.getSourceUri().isEmpty()) {
        sourceUri_ = other.sourceUri_;
        onChanged();
      }
      if (!other.sourceAccessions_.isEmpty()) {
        if (sourceAccessions_.isEmpty()) {
          sourceAccessions_ = other.sourceAccessions_;
          bitField0_ = (bitField0_ & ~0x00000020);
        } else {
          ensureSourceAccessionsIsMutable();
          sourceAccessions_.addAll(other.sourceAccessions_);
        }
        onChanged();
      }
      if (other.getNcbiTaxonId() != 0) {
        setNcbiTaxonId(other.getNcbiTaxonId());
      }
      onChanged();
      return this;
    }

    public final boolean isInitialized() {
      return true;
    }

    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      com.google.genomics.v1.Reference parsedMessage = null;
      try {
        parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        parsedMessage = (com.google.genomics.v1.Reference) e.getUnfinishedMessage();
        throw e.unwrapIOException();
      } finally {
        if (parsedMessage != null) {
          mergeFrom(parsedMessage);
        }
      }
      return this;
    }
    private int bitField0_;

    private java.lang.Object id_ = "";
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     */
    public java.lang.String getId() {
      java.lang.Object ref = id_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        id_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     */
    public com.google.protobuf.ByteString
        getIdBytes() {
      java.lang.Object ref = id_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        id_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     */
    public Builder setId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      id_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     */
    public Builder clearId() {
      
      id_ = getDefaultInstance().getId();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     */
    public Builder setIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      id_ = value;
      onChanged();
      return this;
    }

    private long length_ ;
    /**
     * <pre>
     * The length of this reference's sequence.
     * </pre>
     *
     * <code>int64 length = 2;</code>
     */
    public long getLength() {
      return length_;
    }
    /**
     * <pre>
     * The length of this reference's sequence.
     * </pre>
     *
     * <code>int64 length = 2;</code>
     */
    public Builder setLength(long value) {
      
      length_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The length of this reference's sequence.
     * </pre>
     *
     * <code>int64 length = 2;</code>
     */
    public Builder clearLength() {
      
      length_ = 0L;
      onChanged();
      return this;
    }

    private java.lang.Object md5Checksum_ = "";
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     */
    public java.lang.String getMd5Checksum() {
      java.lang.Object ref = md5Checksum_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        md5Checksum_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     */
    public com.google.protobuf.ByteString
        getMd5ChecksumBytes() {
      java.lang.Object ref = md5Checksum_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        md5Checksum_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     */
    public Builder setMd5Checksum(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      md5Checksum_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     */
    public Builder clearMd5Checksum() {
      
      md5Checksum_ = getDefaultInstance().getMd5Checksum();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     */
    public Builder setMd5ChecksumBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      md5Checksum_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object name_ = "";
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     */
    public java.lang.String getName() {
      java.lang.Object ref = name_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        name_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     */
    public com.google.protobuf.ByteString
        getNameBytes() {
      java.lang.Object ref = name_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        name_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     */
    public Builder setName(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      name_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     */
    public Builder clearName() {
      
      name_ = getDefaultInstance().getName();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     */
    public Builder setNameBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      name_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object sourceUri_ = "";
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     */
    public java.lang.String getSourceUri() {
      java.lang.Object ref = sourceUri_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        sourceUri_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     */
    public com.google.protobuf.ByteString
        getSourceUriBytes() {
      java.lang.Object ref = sourceUri_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        sourceUri_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     */
    public Builder setSourceUri(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      sourceUri_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     */
    public Builder clearSourceUri() {
      
      sourceUri_ = getDefaultInstance().getSourceUri();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     */
    public Builder setSourceUriBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      sourceUri_ = value;
      onChanged();
      return this;
    }

    private com.google.protobuf.LazyStringList sourceAccessions_ = com.google.protobuf.LazyStringArrayList.EMPTY;
    private void ensureSourceAccessionsIsMutable() {
      if (!((bitField0_ & 0x00000020) == 0x00000020)) {
        sourceAccessions_ = new com.google.protobuf.LazyStringArrayList(sourceAccessions_);
        bitField0_ |= 0x00000020;
       }
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public com.google.protobuf.ProtocolStringList
        getSourceAccessionsList() {
      return sourceAccessions_.getUnmodifiableView();
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public int getSourceAccessionsCount() {
      return sourceAccessions_.size();
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public java.lang.String getSourceAccessions(int index) {
      return sourceAccessions_.get(index);
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public com.google.protobuf.ByteString
        getSourceAccessionsBytes(int index) {
      return sourceAccessions_.getByteString(index);
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public Builder setSourceAccessions(
        int index, java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  ensureSourceAccessionsIsMutable();
      sourceAccessions_.set(index, value);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public Builder addSourceAccessions(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  ensureSourceAccessionsIsMutable();
      sourceAccessions_.add(value);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public Builder addAllSourceAccessions(
        java.lang.Iterable<java.lang.String> values) {
      ensureSourceAccessionsIsMutable();
      com.google.protobuf.AbstractMessageLite.Builder.addAll(
          values, sourceAccessions_);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public Builder clearSourceAccessions() {
      sourceAccessions_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      bitField0_ = (bitField0_ & ~0x00000020);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     */
    public Builder addSourceAccessionsBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      ensureSourceAccessionsIsMutable();
      sourceAccessions_.add(value);
      onChanged();
      return this;
    }

    private int ncbiTaxonId_ ;
    /**
     * <pre>
     * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
     * </pre>
     *
     * <code>int32 ncbi_taxon_id = 7;</code>
     */
    public int getNcbiTaxonId() {
      return ncbiTaxonId_;
    }
    /**
     * <pre>
     * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
     * </pre>
     *
     * <code>int32 ncbi_taxon_id = 7;</code>
     */
    public Builder setNcbiTaxonId(int value) {
      
      ncbiTaxonId_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
     * </pre>
     *
     * <code>int32 ncbi_taxon_id = 7;</code>
     */
    public Builder clearNcbiTaxonId() {
      
      ncbiTaxonId_ = 0;
      onChanged();
      return this;
    }
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return this;
    }

    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return this;
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.Reference)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.Reference)
  private static final com.google.genomics.v1.Reference DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.Reference();
  }

  public static com.google.genomics.v1.Reference getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<Reference>
      PARSER = new com.google.protobuf.AbstractParser<Reference>() {
    public Reference parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
        return new Reference(input, extensionRegistry);
    }
  };

  public static com.google.protobuf.Parser<Reference> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<Reference> getParserForType() {
    return PARSER;
  }

  public com.google.genomics.v1.Reference getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

