// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/annotations.proto

package com.google.genomics.v1;

public interface VariantAnnotationOrBuilder extends
    // @@protoc_insertion_point(interface_extends:google.genomics.v1.VariantAnnotation)
    com.google.protobuf.MessageOrBuilder {

  /**
   * <pre>
   * Type has been adapted from ClinVar's list of variant types.
   * </pre>
   *
   * <code>.google.genomics.v1.VariantAnnotation.Type type = 1;</code>
   */
  int getTypeValue();
  /**
   * <pre>
   * Type has been adapted from ClinVar's list of variant types.
   * </pre>
   *
   * <code>.google.genomics.v1.VariantAnnotation.Type type = 1;</code>
   */
  com.google.genomics.v1.VariantAnnotation.Type getType();

  /**
   * <pre>
   * Effect of the variant on the coding sequence.
   * </pre>
   *
   * <code>.google.genomics.v1.VariantAnnotation.Effect effect = 2;</code>
   */
  int getEffectValue();
  /**
   * <pre>
   * Effect of the variant on the coding sequence.
   * </pre>
   *
   * <code>.google.genomics.v1.VariantAnnotation.Effect effect = 2;</code>
   */
  com.google.genomics.v1.VariantAnnotation.Effect getEffect();

  /**
   * <pre>
   * The alternate allele for this variant. If multiple alternate alleles
   * exist at this location, create a separate variant for each one, as they
   * may represent distinct conditions.
   * </pre>
   *
   * <code>string alternate_bases = 3;</code>
   */
  java.lang.String getAlternateBases();
  /**
   * <pre>
   * The alternate allele for this variant. If multiple alternate alleles
   * exist at this location, create a separate variant for each one, as they
   * may represent distinct conditions.
   * </pre>
   *
   * <code>string alternate_bases = 3;</code>
   */
  com.google.protobuf.ByteString
      getAlternateBasesBytes();

  /**
   * <pre>
   * Google annotation ID of the gene affected by this variant. This should
   * be provided when the variant is created.
   * </pre>
   *
   * <code>string gene_id = 4;</code>
   */
  java.lang.String getGeneId();
  /**
   * <pre>
   * Google annotation ID of the gene affected by this variant. This should
   * be provided when the variant is created.
   * </pre>
   *
   * <code>string gene_id = 4;</code>
   */
  com.google.protobuf.ByteString
      getGeneIdBytes();

  /**
   * <pre>
   * Google annotation IDs of the transcripts affected by this variant. These
   * should be provided when the variant is created.
   * </pre>
   *
   * <code>repeated string transcript_ids = 5;</code>
   */
  java.util.List<java.lang.String>
      getTranscriptIdsList();
  /**
   * <pre>
   * Google annotation IDs of the transcripts affected by this variant. These
   * should be provided when the variant is created.
   * </pre>
   *
   * <code>repeated string transcript_ids = 5;</code>
   */
  int getTranscriptIdsCount();
  /**
   * <pre>
   * Google annotation IDs of the transcripts affected by this variant. These
   * should be provided when the variant is created.
   * </pre>
   *
   * <code>repeated string transcript_ids = 5;</code>
   */
  java.lang.String getTranscriptIds(int index);
  /**
   * <pre>
   * Google annotation IDs of the transcripts affected by this variant. These
   * should be provided when the variant is created.
   * </pre>
   *
   * <code>repeated string transcript_ids = 5;</code>
   */
  com.google.protobuf.ByteString
      getTranscriptIdsBytes(int index);

  /**
   * <pre>
   * The set of conditions associated with this variant.
   * A condition describes the way a variant influences human health.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantAnnotation.ClinicalCondition conditions = 6;</code>
   */
  java.util.List<com.google.genomics.v1.VariantAnnotation.ClinicalCondition> 
      getConditionsList();
  /**
   * <pre>
   * The set of conditions associated with this variant.
   * A condition describes the way a variant influences human health.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantAnnotation.ClinicalCondition conditions = 6;</code>
   */
  com.google.genomics.v1.VariantAnnotation.ClinicalCondition getConditions(int index);
  /**
   * <pre>
   * The set of conditions associated with this variant.
   * A condition describes the way a variant influences human health.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantAnnotation.ClinicalCondition conditions = 6;</code>
   */
  int getConditionsCount();
  /**
   * <pre>
   * The set of conditions associated with this variant.
   * A condition describes the way a variant influences human health.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantAnnotation.ClinicalCondition conditions = 6;</code>
   */
  java.util.List<? extends com.google.genomics.v1.VariantAnnotation.ClinicalConditionOrBuilder> 
      getConditionsOrBuilderList();
  /**
   * <pre>
   * The set of conditions associated with this variant.
   * A condition describes the way a variant influences human health.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantAnnotation.ClinicalCondition conditions = 6;</code>
   */
  com.google.genomics.v1.VariantAnnotation.ClinicalConditionOrBuilder getConditionsOrBuilder(
      int index);

  /**
   * <pre>
   * Describes the clinical significance of a variant.
   * It is adapted from the ClinVar controlled vocabulary for clinical
   * significance described at:
   * http://www.ncbi.nlm.nih.gov/clinvar/docs/clinsig/
   * </pre>
   *
   * <code>.google.genomics.v1.VariantAnnotation.ClinicalSignificance clinical_significance = 7;</code>
   */
  int getClinicalSignificanceValue();
  /**
   * <pre>
   * Describes the clinical significance of a variant.
   * It is adapted from the ClinVar controlled vocabulary for clinical
   * significance described at:
   * http://www.ncbi.nlm.nih.gov/clinvar/docs/clinsig/
   * </pre>
   *
   * <code>.google.genomics.v1.VariantAnnotation.ClinicalSignificance clinical_significance = 7;</code>
   */
  com.google.genomics.v1.VariantAnnotation.ClinicalSignificance getClinicalSignificance();
}
